lightning/src/util/persist.rs

```diff
@@ -194,3 +199,800 @@ where
 	}
 	Ok(res)
 }
+
+enum KVStoreUpdatingPersisterError {
+	/// The monitor name was improperly formatted.
+	BadMonitorName(String, String),
+	/// The monitor could not be decoded.
+	MonitorDecodeFailed(DecodeError, String),
+	/// The update could not be decoded.
+	UpdateDecodeFailed(DecodeError, String),
+	/// Storage could not be read.
+	StorageReadFailed(io::Error, String),
+	/// An update could not be applied to a monitor.
+	UpdateFailed(String, String),
+}
+
+impl From<KVStoreUpdatingPersisterError> for io::Error {
+	fn from(value: KVStoreUpdatingPersisterError) -> Self {
+		match value {
+			KVStoreUpdatingPersisterError::BadMonitorName(reason, context) => io::Error::new(
+				io::ErrorKind::InvalidInput,
+				format!("{reason}, context: {context}'"),
+			),
+			KVStoreUpdatingPersisterError::MonitorDecodeFailed(reason, context) => io::Error::new(
+				io::ErrorKind::InvalidData,
+				format!("{reason}, context: {context:?}'"),
+			),
+			KVStoreUpdatingPersisterError::UpdateDecodeFailed(reason, context) => io::Error::new(
+				io::ErrorKind::InvalidData,
+				format!("{reason}, context: {context:?}'"),
+			),
+			KVStoreUpdatingPersisterError::StorageReadFailed(reason, context) => {
+				io::Error::new(io::ErrorKind::Other, format!("{reason}, context: {context:?}'"))
+			}
+			KVStoreUpdatingPersisterError::UpdateFailed(reason, context) => {
+				io::Error::new(io::ErrorKind::InvalidData, format!("{reason}, context: {context}'"))
+			}
+		}
+	}
+}
+
+/// A struct representing a name for a monitor.
+#[derive(Clone, Debug)]
+pub struct MonitorName(String);
+
+impl MonitorName {
+    /// The key to store this monitor with.
+    fn storage_key(&self) -> String {
+        self.0.clone()
+    }
+}
+
+impl TryFrom<MonitorName> for OutPoint {
+    type Error = std::io::Error;
+
+    fn try_from(value: MonitorName) -> Result<Self, io::Error> {
+        let (txid_hex, index) = value.0.split_once('_').ok_or_else(|| {
+            KVStoreUpdatingPersisterError::BadMonitorName(
+                "no underscore".to_string(),
+                value.0.clone(),
+            )
+        })?;
+        let index = index.parse().map_err(|e| {
+            KVStoreUpdatingPersisterError::BadMonitorName(
+                format!("bad index value, caused by {e}"),
+                value.0.clone(),
+            )
+        })?;
+        let txid = Txid::from_hex(txid_hex).map_err(|e| {
+            KVStoreUpdatingPersisterError::BadMonitorName(
+                format!("bad txid, caused by: {e}"),
+                value.0.clone(),
+            )
+        })?;
+        Ok(OutPoint { txid, index })
+    }
+}
+
+impl From<OutPoint> for MonitorName {
+    fn from(value: OutPoint) -> Self {
+        MonitorName(format!("{}_{}", value.txid.to_hex(), value.index))
+    }
+}
+
+/// A struct representing a name for an update.
+#[derive(Clone, Debug, PartialEq, PartialOrd, Eq, Ord)]
+pub struct UpdateName(String);
```

Comment:
Ack on the borrows.

I created these structs as [newtypes](https://doc.rust-lang.org/rust-by-example/generics/new_types.html), initially, so that I could more easily keep things straight at design time, like on some of the return types when I'm listing things. Then when I had a couple of conversions, the newtype struct was a sensible place to organize and access them.

So, I'd prefer to keep them, as I think they make the code more explicit. That said, if it's bothersome, I'm content to consider the newtypes "scaffolding" and refactor it to just be helper functions.

I do think putting them on the CM/CMU is probably not the way I'd encapsulate, as this "txidhex_idx" identifier is specific to storage keys and doesn't really belong in the Lightning domain. But as usual I am not strongly opinionated here.

Reply:
I changed this so that these types are now in a submodule, which makes direct instantiation impossible, and added constructors with validation.

Reply:
Well, okay. The reason I did the internal submodule was to prevent future authors (including myself in 6 months) from bypassing the validation by accident. The inner `String` is a private field so you can't directly construct this:

```
error[E0423]: cannot initialize a tuple struct which contains private fields
   --> lightning/src/util/persist.rs:371:13
    |
371 |         let key = MonitorName(key.to_string());
    |                   ^^^^^^^^^^^
    |
    ```
I guess that's overkill? We can just rely on reviews to make sure the validation is not bypassed.